Sha3/src/Keccak.java

```diff
@@ -350,53 +376,87 @@ public byte getAlgorithm() {
                 return ALG_SHA3_512;
         }
     }
+
+    // Set shake return length
+    public void setShakeDigestLength(short length)
+    {
+        mdlen = length;
+    }
     
-    // get sha3 instance
-    public static Sha3 getInstance(byte algorithm) throws CryptoException {
+    // get Keccak instance
+    public static Keccak getInstance(byte algorithm) throws CryptoException {
         switch (algorithm) {
             case ALG_KECCAK_224:
                 pad = 0x01;
+                mdlen = (short) 28;
+                rsiz = (short) 144;
+                break;
             case ALG_SHA3_224:
                 //not supported by MessageDigest
+                pad = 0x06;
                 mdlen = (short) 28;
                 rsiz = (short) 144;
                 break;
-
             case ALG_KECCAK_256:
                 pad = 0x01;
+                mdlen = (short) 32;
+                rsiz = (short) 136;
+                break;
             case ALG_SHA3_256:
             case ALG_SHA_256:
+                pad = 0x06;
                 mdlen = (short) 32;
                 rsiz = (short) 136;
                 break;
             case ALG_KECCAK_384:
                 pad = 0x01;
+                mdlen = (short) 48;
+                rsiz = (short) 104;
+                break;
             case ALG_SHA3_384:
             case ALG_SHA_384:
+                pad = 0x06;
                 mdlen = (short) 48;
                 rsiz = (short) 104;
                 break;
             case ALG_KECCAK_512:
                 pad = 0x01;
+                mdlen = (short) 64;
+                rsiz = (short) 72;
+                break;
             case ALG_SHA3_512:
             case ALG_SHA_512:
+                pad = 0x06;
                 mdlen = (short) 64;
                 rsiz = (short) 72;
                 break;
+            case ALG_SHAKE_128:
+                pad = 0x1F;
+                //Default 32, length can be set by calling setShakeDigestLength()
+                mdlen = (short) 32;
+                rsiz = (short) 168;
+                break;
+            case ALG_SHAKE_256:
+                pad = 0x1F;
+                //Default 64, length can be set by calling setShakeDigestLength()
+                mdlen = (short) 64;
+                rsiz = (short) 136;
+                break;
             default:
                 throw new CryptoException(CryptoException.NO_SUCH_ALGORITHM);
         }
         pt = 0;
         
         if (m_instance == null) {
-            m_instance = new Sha3();
+            m_instance = new Keccak();
         }
         return m_instance;
     }
     
     @Override
     public byte getLength() {
-        return mdlen;
+        //Cast to byte is problematic for Shake
```

Comment:
Can you pls elaborate why this cast is problematic and what it means. Right now, cast is there and note that it is problematic but it is unclear if it is fixed (missing cast was the problem) or something else.

Reply:
The main issue is that the getLength function in the abstract MessageDigest class is of return type byte, where the length of shake can be set to max ~32000 (signed short), but since it is overridden, its not possible to change its type. Alternativly the abstraction could be removed (this is what I done on my own project), otherwise it has to be cast and if the length is set to larger than 127 it will cast and show incorrect results.

Reply:
ok, clear - can you please add this whole reasoning into the code comment?

Reply:
Yes, I will do this tomorow